socorro/external/pubsub/crashqueue.py

```diff
@@ -0,0 +1,150 @@
+        reference_value_from='resource.pubsub'
+    )
+    required_config.add_option(
+        'priority_topic_name',
+        doc='The topic name for the priority queue.',
+        reference_value_from='resource.pubsub'
+    )
+    required_config.add_option(
+        'priority_subscription_name',
+        doc='The subscription name for the priority queue.',
+        reference_value_from='resource.pubsub'
+    )
+    required_config.add_option(
+        'reprocessing_topic_name',
+        doc='The topic name for the reprocessing queue.',
+        reference_value_from='resource.pubsub'
+    )
+    required_config.add_option(
+        'reprocessing_subscription_name',
+        doc='The subscription name for the reprocessing queue.',
+        reference_value_from='resource.pubsub'
+    )
+
+    def __init__(self, config, namespace='', quit_check_callback=None):
+        self.config = config
+        self.quit_check_callback = quit_check_callback
+
+        if 'PUBSUB_EMULATOR_HOST' in os.environ:
+            self.subscriber = pubsub_v1.SubscriberClient()
+        else:
+            self.subscriber = pubsub_v1.SubscriberClient.from_service_account_file(
+                config.service_account_file
+            )
+        self.standard_path = self.subscriber.subscription_path(
+            config.project_id, config.standard_subscription_name
+        )
+        self.priority_path = self.subscriber.subscription_path(
+            config.project_id, config.priority_subscription_name
+        )
+        self.reprocessing_path = self.subscriber.subscription_path(
+            config.project_id, config.reprocessing_subscription_name
+        )
+
+        # NOTE(willkg): This will fail if one of the subscription names don't
+        # exist
+        self.subscriber.get_subscription(self.standard_path)
+        self.subscriber.get_subscription(self.priority_path)
+        self.subscriber.get_subscription(self.reprocessing_path)
+
+    def ack_crash(self, sub_path, ack_id):
+        self.subscriber.acknowledge(sub_path, [ack_id])
+        logger.debug('ack %s from %s', ack_id, sub_path)
+
+    def close(self):
+        pass
+
+    def __iter__(self):
+        """Return iterator over crash ids from Pub/Sub.
+
+        Each returned crash is a ``(crash_id, {kwargs})`` tuple with
+        ``finished_func`` as the only key in ``kwargs``. The caller should call
+        ``finished_func`` when it's done processing the crash.
+
+        This does a single pass through the queues and pulls at most 10 from
+        each queue. It tries twice to pull from the priority queue since it's
+        a priority queue.
+
+        """
+        queues = [
+            self.priority_path,
+            self.standard_path,
+            self.reprocessing_path,
+            self.priority_path
+        ]
+
+        for queue_path in queues:
+            response = self.subscriber.pull(
+                queue_path,
+                max_messages=1,
```

Comment:
This goes through all the queues and pulls at most one crash id. That way we're not sitting on a whole bunch of crash ids with ticking ack deadlines.